Add ManagedObjectReference and more events supported

embano1 opened the pull request:
- Changes:
  - Many important vCenter events supported now
  - Split event handling in separate pkg
  - JSON key in outbound Message is now `ManagedObjectReference` for functions to check
  - Robust vCenter client connection/error handling
  - Clean shutdown via `os.Signal`
  - Fixed a bug where `-insecure` was never populated
  - Improved error handling and logging (removed panics)
  - More comments in the code
  - Dockerfiles updated
  - `dep ensure` was out of sync
  - README fixes and updates reflecting new events supported
  - New `docs` section for examples
  - Bumped Docker image version for the connector to `0.4` (until official
OpenFaaS builds are used)
  - Updated external examples (gotagfn, pytagfn) to use new JSON key

- Tested with:
  - vCenter 6.5
  - Kubernetes: kindest/node:v1.13.3 (kind version 0.2.0-alpha)
  - OpenFaaS: faas-netes (Commit 07bca00)
  - Python demo function [pytagfn](https://github.com/embano1/pytagfn) (Commit 99e2eea)

- Improvements for further PRs:
  - Migrate to latest openfaas-sdk version
  - Add code examples to repo instead personal repo
  - Switch to official builds in YAML
  - When govmomi ships govc/vcsim Docker images, use them

Signed-off-by: Michael Gasch <embano1@live.com>

ivanayov commented:
Hi @embano1 
Can you please add a "How has this been tested" section?
Looks good

embano1 commented:
Thx @ivanayov 

Actually, I found some more issues, e.g. the "insecure" flag actually was never used. I've done several changes and need to do more testing.

How would you like me to update this PR in terms of commit? Squash them all into one commit and force push?

ivanayov commented:
@embano1 if they are fixes and refactoring of the same change, then should better go in a single commit

alexellis commented on main.go:
```diff
@@ -25,14 +25,10 @@ func main() {
-	flag.StringVar(&vcenterURL, "vcenter-url", "", "URL for vcenter user:password@domain.com:port")
-	flag.Bool("insecure", true, "use an insecure connection")
+	flag.StringVar(&vcenterURL, "vcenter-url", "http://user:pass@127.0.0.1:8989/sdk", "URL for vCenter")
+	flag.Bool("insecure", false, "use an insecure connection to vCenter (default false)")
```
Please make sure to update all the examples if we make this breaking change.

embano1 replied:
Currently we don't have examples in this repo, just externally which makes it hard to track (besides my examples of course). Also, the `insecure` flag was always `false` as this was a bug in the implementation (i.e. not using BoolVar but Bool in the old code). So the fix basically is the old behaviour (`false`).

embano1 replied:
Btw: this is fixed in an upcoming (pending) commit.

alexellis commented on main.go:
```diff
@@ -149,22 +149,32 @@ func handleEvent(event vtypes.BaseEvent, m *event.Manager) (string, string, erro
+		return "", "", errors.Wrap(err, "error marshaling outboundevent")
+	}
+
+	log.Println(string(message))
```
Can we do the print one level up from here and only if a verbose env-var is set?

embano1 replied:
Which format do you prefer for the env, e.g. `VCVERBOSE`?

embano1 replied:
will fix in a follow-up PR if ok

alexellis commented on main.go:
```diff
@@ -149,22 +149,32 @@ func handleEvent(event vtypes.BaseEvent, m *event.Manager) (string, string, erro
-	CreatedTime time.Time `json:"createdTime,omitempty"`
+	UserName    string                         `json:"userName,omitempty"`
+	CreatedTime time.Time                      `json:"createdTime,omitempty"`
+	MoRef       *vtypes.ManagedObjectReference `json:"moref,omitempty"`
```
Should we go for `ManagedObjectReference`?

embano1 replied:
Reason I choose a pointer is that if *ManagedObjectReference is nil, it won't be marshaled. If it's a struct, it will be marshaled, bypassing "omitempty" (https://play.golang.org/p/z6cwin69PDC). Thus, on the receiver side it's easier to unmarshal and check whether the field exists/is nil (depending on the language and JSON implementation) then dealing with `types.ManagedObjectReference{}` being empty.

alexellis replied:
Sorry I didn't explain that well. Should we rename `MoRef` to the longer variant?

embano1 replied:
Will change to:

```go
// OutboundEvent is the JSON object sent to subscribed functions
// If the ManagedObjectReference for an event cannot be retrieved, it will be nil and thus not marshaled into the JSON OutboundEvent
// It's the receivers responsibility to check whether managedObjectReference key is present in the JSON message payload
type OutboundEvent struct {
	Topic    string `json:"topic,omitempty"`
	Category string `json:"category,omitempty"`

	UserName               string                         `json:"userName,omitempty"`
	CreatedTime            time.Time                      `json:"createdTime,omitempty"`
	ManagedObjectReference *vtypes.ManagedObjectReference `json:"managedObjectReference,omitempty"`
}
```

embano1 replied:
- [ ] Update Go and Python examples on changed JSON key for MoRef

alexellis commented on main.go:
```diff
@@ -25,14 +25,10 @@ func main() {
+	flag.Bool("insecure", false, "use an insecure connection to vCenter (default false)")
 	flag.Parse()
 
-	if len(vcenterURL) == 0 {
```
Why was this removed?

embano1 replied:
flag.StringVar sets a default value for it, unless overwritten (to "") by the user, which is unlikely. Even if the user overwrites it to "" it would be caught in the connection logic in newVCenterClient by govmomi.

alexellis commented on main.go:
```diff
@@ -149,31 +147,42 @@ func handleEvent(event vtypes.BaseEvent, m *event.Manager) (string, string, erro
+	u, err := soap.ParseURL(vcenterURL)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "error parsing URL")
```
💯

alexellis commented:
Is this still WIP?

embano1 commented:
PR updated with changes outlined in the commit message.

embano1 commented on main.go:
```diff
@@ -22,25 +23,25 @@ import (
+	flag.StringVar(&gatewayURL, "gateway", "http://127.0.0.1:8080", "URL for OpenFaaS gateway")
+	flag.StringVar(&vcenterURL, "vcenter", "http://127.0.0.1:8989/sdk", "URL for vCenter")
+	flag.StringVar(&vcUser, "vc-user", "", "User to connect to vCenter")
+	flag.StringVar(&vcPass, "vc-pass", "", "Password to connect to vCenter")
```
will fix in a follow-up PR with improved secrets handling if ok

alexellis commented on Dockerfile:
```diff
@@ -3,7 +3,7 @@ RUN mkdir -p /go/src/github.com/openfaas-incubator/vcenter-connector
 
 COPY vendor     vendor
-COPY main.go    .
+COPY *.go    ./
```
I'd rather you didn't change this. Specifying individual files allows for better caching.

alexellis commented on events.go:
```diff
@@ -0,0 +1,196 @@
+	// Get the underlying concrete base event, e.g. VmEvent
+	// vSphere Web Service API Reference 6.7
+	// https://code.vmware.com/apis/358/vsphere#/doc/vim.event.Event.html
+	switch baseEvent := event.(type) {
```
This needs to be extracted to a new function and unit tested. Single Responsibility Principle.

alexellis commented on events.go:
```diff
@@ -0,0 +1,196 @@
+// OutboundEvent is the JSON object sent to subscribed functions
+// If the ManagedObjectReference for an event cannot be retrieved, it will be nil and thus not marshaled into the JSON OutboundEvent
+// It's the receivers responsibility to check whether managedObjectReference key is present in the JSON message payload
+type OutboundEvent struct {
```
I'm not sure that having multiple files makes sense when they are all in the same package.

If you want to split them out then consider also breaking out a package / sub-folder.

alexellis commented on README.md:
```diff
@@ -10,14 +10,24 @@ This project uses the [OpenFaaS Connector SDK](https://github.com/openfaas-incub
+
+For further details and naming see the [vSphere Web Services API](https://code.vmware.com/apis/358/vsphere#/doc/vim.event.Event.html) documentation.
 
 ## Example: vCenter Tagging Function
```
Let's break this out into a separate markdown file?

./docs/example.md

For instance.

alexellis commented on events.go:
```diff
@@ -0,0 +1,196 @@
+	return topic, string(message), nil
+}
+
+// convertToTopic converts an event type to an OpenFaaS subscriber topic, e.g. "VmPoweredOnEvent" to "vm.powered.on"
```
Add at least one unit test for this.

alexellis replied:
https://blog.alexellis.io/golang-writing-unit-tests/

embano1 commented:
@alexellis thx for your thorough review! What I changed based on your feedback:

- fixed the Dockerfiles
- example is now in `docs`
- moved event logic into a separate package (events.go)
- added a function `getmoref` to break out the `switch` statements for the events, returning `*vtypes.ManagedObjectReference`
- added a test for `getmoref()` in `events.go` to check for valid/supported events

alexellis commented:
Next step - let's see the password read from a Kubernetes secret file?

Alex
